zebra-state/src/sled_state.rs

```diff
@@ -66,38 +67,39 @@ impl FinalizedState {
         let prev_hash = queued_block.block.header.previous_block_hash;
         self.queued_by_prev_hash.insert(prev_hash, queued_block);
 
-        // Cloning means the closure doesn't hold a borrow of &self,
-        // conflicting with mutable access in the loop below.
-        let hash_by_height = self.hash_by_height.clone();
-        let tip_hash = || {
-            read_tip(&hash_by_height)
-                .expect("inability to look up tip is unrecoverable")
-                .map(|(_height, hash)| hash)
-                .unwrap_or(block::Hash([0; 32]))
-        };
-
-        while let Some(queued_block) = self.queued_by_prev_hash.remove(&tip_hash()) {
+        while let Some(queued_block) = self.queued_by_prev_hash.remove(&self.finalized_tip_hash()) {
             self.commit_finalized(queued_block)
         }
     }
 
-    /// Commit a finalized block to the state.
-    ///
-    /// It's the caller's responsibility to ensure that blocks are committed in
-    /// order. This function is called by [`process_queue`], which ensures order.
-    /// It is intentionally not exposed as part of the public API of the
-    /// [`FinalizedState`].
-    fn commit_finalized(&mut self, queued_block: QueuedBlock) {
-        let QueuedBlock { block, rsp_tx } = queued_block;
+    pub fn finalized_tip_hash(&self) -> block::Hash {
+        read_tip(&self.hash_by_height)
+            .expect("inability to look up tip is unrecoverable")
+            .map(|(_, hash)| hash)
+            .unwrap_or(block::Hash([0; 32]))
+    }
+
+    pub fn finalized_tip_height(&self) -> block::Height {
+        read_tip(&self.hash_by_height)
+            .expect("inability to look up tip is unrecoverable")
+            .map(|(height, _)| height)
+            .unwrap_or(block::Height(0))
```

Comment:
The empty state value is inconsistent with `finalized_tip_hash`, the height of the genesis previous block is undefined.

Is there a test for empty state?
If there is, that test is also incorrect.
If not, please add one.

Let's also make sure we explain the empty state behavior in the function comment.

```suggestion
    pub fn finalized_tip_height(&self) -> Option<block::Height> {
        read_tip(&self.hash_by_height)
            .expect("inability to look up tip is unrecoverable")
            .map(|(height, _)| height)
            // if the state is empty, the genesis previous block height is undefined
            .unwrap_or(None)
```

Reply:
Un-resolved, because this bug should also have a working test.

Reply:
I added a todo on the follow up PR for the tests